### ISSUE TITLE
Support tailwind v4

### ISSUE THREAD

**Issue by heddendorp** [angular/angular-cli#28938]
### Command

build

### Description

Tailwind v4 is changing the installation and configuration. It would be great to update the cli to make sure v4 is supported.

### Describe the solution you'd like

The cli should support the new postcss plugin https://tailwindcss.com/docs/v4-beta#installing-with-post-css
Also detection probably has to be changed as v4 no longer uses js to configure tailwind but instead css variables https://tailwindcss.com/docs/v4-beta#installing-with-post-css
I guess the best detection at this point is having tailwindcss with a version >4 installed and ideally the cli also makes sure @tailwindcss/postcss is installed.

### Describe alternatives you've considered

Tailwind also offers a vite plugin, but to my understanding, postcss should be easier to support

**Comment by clydin** [comment]
Custom postcss configurations are possible with the `application` build system (default for all new projects).  The `@tailwindcss/postcss` plugin can be used directly by creating a `postcss.config.json`/``.postcssrc.json` file and including the plugin.

As tailwindcss v4 moves towards a stable release, we are also evaluating direct inclusion of the optimized pipeline present in the Vite plugin to avoid the need for postcss if preferred.

**Comment by pumano** [comment]
@clydin looks like `vite` plugin should be more performant than `postcss`

**Comment by dcheglakov** [comment]
@clydin I’ve tried following your approach with the `postcss.config.json` file and enabling Tailwind CSS via PostCSS like this:
```json
{
  "plugins": {
    "@tailwindcss/postcss": {}
  }
}
```
for my Angular 16 project. Unfortunately, it didn’t resolve the issue for me. Maybe there’s something specific to Angular 16 that I might be missing? Thank you.

**Comment by clydin** [comment]
Custom postcss configuration is only supported with the `application` build system in v17.2 or higher.

**Comment by JoepKockelkorn** [comment]
I can confirm the `postcss.config.json` approach works. I did have to install tailwind and its postcss plugin using `--legacy-peer-deps` due to conflicts.

**Comment by Char2sGu** [comment]
> Custom postcss configurations are possible with the `application` build system (default for all new projects).

Should this be documented? I would never know about this if I didn't accidentally clicked into this issue. @clydin

**Comment by sheikalthaf** [comment]
![Image](https://github.com/user-attachments/assets/11982f69-10e9-4fc7-b5a8-8fd1be1f30f6)

Tailwind v4 is getting release next week and we need to support it from angular is good.

**Comment by kyjus25** [comment]
<img width="589" alt="Image" src="https://github.com/user-attachments/assets/3ddc6d4c-eab7-49c1-b00d-f940a1a0a74c" />

Looks like Jan 21 might be the day!

**Comment by e-oz** [comment]
And it's released!

**Comment by e-oz** [comment]
Their [installation guide for Angular](https://tailwindcss.com/docs/installation/framework-guides/angular) works fine, I was able to make it work in an Angular v19 app following that guide, even with plugins.

It would be nice if we could do it without `.postcssrc.json` though ;)

**Comment by heddendorp** [comment]
> Their [installation guide for Angular](https://tailwindcss.com/docs/installation/framework-guides/angular) works fine, I was able to make it work in an Angular v19 app following that guide, even with plugins.
> 
> It would be nice if we could do it without `.postcssrc.json` though ;)

A small downside here is that the angular cli package complains about using v4

**Comment by Kludgy4** [comment]
Released! Would be nice to not have to use postcss to also use tailwind

**Comment by bboyle** [comment]
this peerDependency is locked to tailwind v2–3: https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/package.json#L84

**Comment by matiascamiletti** [comment]
> this peerDependency is locked to tailwind v2–3: https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/package.json#L84

![Image](https://github.com/user-attachments/assets/6741d2fd-f345-4983-a884-f378d22a99e1)

I have the same error

**Comment by quedicesebas** [comment]
Same here, can't deploy to Firebase

**Comment by cze-dev** [comment]
> this peerDependency is locked to tailwind v2–3: https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/package.json#L84

would be nice if at least the peer dependency could be updated to `^2.0.0 || ^3.0.0 || ^4.0.0` to avoid usind `--force` or `--legacy-peer-deps` in pipelines

**Comment by dr-schopalopp** [comment]
> this peerDependency is locked to tailwind v2–3: https://github.com/angular/angular-cli/blob/main/packages/angular_devkit/build_angular/package.json#L84

would be nice if at least the peer dependency could be updated to `^2.0.0 || ^3.0.0 || ^4.0.0` to avoid usind `--force` or `--legacy-peer-deps` in pipelines

**Comment by rudyhadoux** [comment]
With Vite and a new Angular 19.1 app, it fails too.

**Comment by rudyhadoux** [comment]
It seems the problem is resolved for me, if I add
@reference "tailwindcss" in css/scss files
everywhere there is @apply.

**Comment by bugproof** [comment]
`@reference` support is far from ideal because angular doesn't allow us to provide custom vite config and we cannot create an alias

stylePreprocessorOptions includePaths is not working with `@reference` so for custom themes/styles you need to use paths like `@reference "../../../../../styles/tailwind.css"

**Comment by rudyhadoux** [comment]
@bugproof it works for me only with @reference "tailwindcss";

**Comment by rudyhadoux** [comment]
@sheikalthaf it is for you.

**Comment by EinfachHans** [comment]
@bugproof I'm having the same issue, the "@reference" is quite annoying to put correctly into every style file...

Is this something angular have to fix or is this something for tailwind itself?

**Comment by bugproof** [comment]
@EinfachHans https://github.com/tailwindlabs/tailwindcss/issues/15778#issuecomment-2618664046

Apparently it's possible to create an alias using vite to make it less painful but angular doesn't allow custom vite configs. It's hard to say. I think tailwind team could give us a better solution for this problem. I have no idea. I tried migrating to v4 in one angular project but it turned out to be too painful due to `@reference` issue and usage of `@apply` in many component styles.

**Comment by clydin** [comment]
Unfortunately, the alias suggestion would not be useful in this case since Tailwind performs its own internal stylesheet resolution and its postcss plugin does not currently provide any options to customize its behavior in this regard.

**Comment by EinfachHans** [comment]
Hey @clydin - thanks for your response!

So the only thing that is left from angular side is the detection of tailwind, that should also work without the "old" tailwind config file?

About the `@reference` stuff, as you have a deeper understanding of it and why an alias won't work, could you maybe open an issue at tailwind and explain it there?

**Comment by bugproof** [comment]
@clydin 

Is the guide from the official tailwindcss docs still relevant? You need `.postcssrc.json` and `@tailwindcss/postcss` package

**Comment by clydin** [comment]
Yes, the guide (https://tailwindcss.com/docs/installation/framework-guides/angular) is still relevant and accurate.  It mentions those packages/files.

**Comment by gamoreli** [comment]
Hey guys, I am getting an error when trying to run ng test:

`An unhandled exception occurred: It looks like you're trying to use `tailwindcss` directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install `@tailwindcss/postcss` and update your PostCSS configuration.`

Any workaround?

**Comment by gamoreli** [comment]
> [@gamoreli](https://github.com/gamoreli) Could you share your `.postcssrc.json`?

`{
  "plugins": {
    "@tailwindcss/postcss": {
      "optimize": {
        "minify": true
      }
    }
  }
}
`

**Comment by gamoreli** [comment]
Apparently we had this issue with other developer some weeks ago, and the issue was move to "Not Planned", I didn't get it. #29690

**Comment by geyuqiu** [comment]
as currently only v3 is working: https://v3.tailwindcss.com/docs/guides/angular

**Comment by ashutoshtyagi-xyz** [comment]
I am also facing same issue in angular. I am creating a library in angular and using storybook.
I tried with both tailwind 3 and tailwind 4.
mainly the issue is that angular is not able to trigger tailwind properly to create classes at runtime, but at compile time it's working.
But major issue is that in default setting, only the tailwind default presets are coming, but tailwind is not able to find the classes, as it is not a running application but library using storybook. So now in default settings and configurations, tailwind is not even generating the classes, for that i found one fix that, if i specify the library path in tailwind config it works as like this:  (https://github.com/ashutoshtyagi-xyz/tailwind3-test), there the issue was fixed when i gave the content path in config file, 

But with the tailwind 4, as default there is no config file, but we can include it which i tried like this:
https://github.com/ashutoshtyagi-xyz/tailwind4-test/tree/new-config

when i tried the same with tailwind4 as config file is allowed, i am getting error:

```
info => Starting manager..
info => Starting preview..
info Addon-docs: using MDX3
info => Using implicit CSS loaders
info => Using angular browser target options from "dsx:build"
info => Using angular project with "tsConfig:/Users/atyagi/Documents/Personal/tailwind4-test/projects/dsx/.storybook/tsconfig.json"
=> Failed to build the preview
Error: It looks like you're trying to use `tailwindcss` directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install `@tailwindcss/postcss` and update your PostCSS configuration.
    at ze (./node_modules/.pnpm/tailwindcss@4.1.3/node_modules/tailwindcss/dist/lib.js:36:1984)
    at getStylesConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/tools/webpack/configs/styles.js:96:66)
    at async Promise.all (index 1)
    at async generateWebpackConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:64:22)
    at async generateBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:119:20)
    at async generateI18nBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:70:20)
    at async exports.getWebpackConfig (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/angular-cli-webpack.js:40:35)
    at async Object.webpackFinal (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/framework-preset-angular-cli.js:20:27)
    at async Object.webpack (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/presets/custom-webpack-preset.js:1:3280)
    at async starter (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/index.js:1:22249)
```

there is something from this line of code i think is not able to process the config file properly in tailwind4, that'll be due to [Angular code](https://github.com/angular/angular-cli/blob/807ba251c244f899c99769cc00c444051d3dd033/packages/angular_devkit/build_angular/src/tools/webpack/configs/styles.ts#L82-L98) that only integrates v3:

```
  const tailwindConfigPath = await findTailwindConfigurationFile(root, projectRoot);
  if (tailwindConfigPath) {
    let tailwindPackagePath;
    try {
      tailwindPackagePath = require.resolve('tailwindcss', { paths: [root] });
    } catch {
      …
    }
    if (tailwindPackagePath) {
      extraPostcssPlugins.push(require(tailwindPackagePath)({ config: tailwindConfigPath }));
    }
  }
```

For v4, it should resolve @tailwindcss/postcss.

If any one has any workaround for this that will be useful.
My target is to have a in-company design system using tailwind.

**Comment by ashutoshtyagi-xyz** [comment]
> I am also facing same issue in angular. I am creating a library in angular and using storybook. I tried with both tailwind 3 and tailwind 4. mainly the issue is that angular is not able to trigger tailwind properly to create classes at runtime, but at compile time it's working. But major issue is that in default setting, only the tailwind default presets are coming, but tailwind is not able to find the classes, as it is not a running application but library using storybook. So now in default settings and configurations, tailwind is not even generating the classes, for that i found one fix that, if i specify the library path in tailwind config it works as like this: (https://github.com/ashutoshtyagi-xyz/tailwind3-test), there the issue was fixed when i gave the content path in config file,
> 
> But with the tailwind 4, as default there is no config file, but we can include it which i tried like this: https://github.com/ashutoshtyagi-xyz/tailwind4-test/tree/new-config
> 
> when i tried the same with tailwind4 as config file is allowed, i am getting error:
> 
> ```
> info => Starting manager..
> info => Starting preview..
> info Addon-docs: using MDX3
> info => Using implicit CSS loaders
> info => Using angular browser target options from "dsx:build"
> info => Using angular project with "tsConfig:/Users/atyagi/Documents/Personal/tailwind4-test/projects/dsx/.storybook/tsconfig.json"
> => Failed to build the preview
> Error: It looks like you're trying to use `tailwindcss` directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install `@tailwindcss/postcss` and update your PostCSS configuration.
>     at ze (./node_modules/.pnpm/tailwindcss@4.1.3/node_modules/tailwindcss/dist/lib.js:36:1984)
>     at getStylesConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/tools/webpack/configs/styles.js:96:66)
>     at async Promise.all (index 1)
>     at async generateWebpackConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:64:22)
>     at async generateBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:119:20)
>     at async generateI18nBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:70:20)
>     at async exports.getWebpackConfig (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/angular-cli-webpack.js:40:35)
>     at async Object.webpackFinal (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/framework-preset-angular-cli.js:20:27)
>     at async Object.webpack (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/presets/custom-webpack-preset.js:1:3280)
>     at async starter (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/index.js:1:22249)
> ```
> 
> there is something from this line of code i think is not able to process the config file properly in tailwind4, that'll be due to [Angular code](https://github.com/angular/angular-cli/blob/807ba251c244f899c99769cc00c444051d3dd033/packages/angular_devkit/build_angular/src/tools/webpack/configs/styles.ts#L82-L98) that only integrates v3:
> 
> ```
>   const tailwindConfigPath = await findTailwindConfigurationFile(root, projectRoot);
>   if (tailwindConfigPath) {
>     let tailwindPackagePath;
>     try {
>       tailwindPackagePath = require.resolve('tailwindcss', { paths: [root] });
>     } catch {
>       …
>     }
>     if (tailwindPackagePath) {
>       extraPostcssPlugins.push(require(tailwindPackagePath)({ config: tailwindConfigPath }));
>     }
>   }
> ```
> 
> For v4, it should resolve @tailwindcss/postcss.
> 
> If any one has any workaround for this that will be useful. My target is to have a in-company design system using tailwind, a library that can be integrated in angular projects.

@clydin can you have a look to this, and can reference to me anything where i am wrong.

**Comment by GlennLefevere** [comment]
Hi,
I see in previous comments that tailwind 4 is currently only supported using the `application` build system.
Are there any plan to add support to the `browser` build system?
At my company we are currently heavily relying on the `browser` build system and migrating to the `application` build system would be a massive undertaking.

**Comment by HarelM** [comment]
I have a project with Angular 19 (latest version to this date - 19.3) that uses the "new" builder: `@angular-devkit/build-angular:application`.
My project is using scss and Angular Material.
I can't seem to be able to make tailwindcss 4 work completely, I did get it to work to some extent, but for example the `top-0` class is missing and I'm not sure what I'm doing wrong.

Does anyone has an example with all of the above that works?

Here's my code if any one would like to take a stab at upgrading tailwindcss:
https://github.com/IsraelHikingMap/Site/tree/ef5f54b037592956452a149371b1b8e9c1de26b3/IsraelHiking.Web
It's also worth noting that I have tried the tailwind upgrade script and it failed.

**Comment by jits** [comment]
@HarelM — here are some potentially useful posts I've written about my experiences with Tailwind v4, Angular, and Angular Material:

- https://jits.dev/blog/migrating-angular-tailwind-v4/
- https://jits.dev/blog/tailwind-v4-play-nice-with-angular-material/

**Comment by HarelM** [comment]
Thanks @jits!
I followed your article but I was not able to use the upgrade tool, I'm getting the following errors:
```
Could not determine configuration file for: `./tailwind.tmp.css` 
│   Update your stylesheet to use `@config` to specify the correct configuration file explicitly and then run the 
│   upgrade tool again. 

│ ↳ Could not determine configuration file for: `./node_modules/tailwindcss/base.css` 
│   Update your stylesheet to use `@config` to specify the correct configuration file explicitly and then run the 
│   upgrade tool again.
```
What am I doing wrong?

**Comment by jits** [comment]
@HarelM — let's take this conversation elsewhere (to avoid pinging everyone watching this issue). I'll do my best to help you out; could you set up an issue on your repo and tag me?

**Comment by martonsimon** [comment]
To use tailwind v4 for ng test:
**File**: _node_modules\@angular-devkit\build-angular\src\tools\webpack\configs\styles.js_
Inside the getStylesConfig(), change this:

```
if (tailwindPackagePath) {
  extraPostcssPlugins.push(require(tailwindPackagePath)({ config: tailwindConfigPath }));
}
```
to this:
```
if (tailwindPackagePath) {
  extraPostcssPlugins.push(require('@tailwindcss/postcss')({ config: tailwindConfigPath }));
}
```

**Comment by kyjus25** [comment]
Can confirm. Tailwind v4 _mostly_ works, but @apply inside component styles do not (they do in v3). CLI throws an error that the style doesn't exist

**Comment by derbohmannatwork** [comment]
@martonsimon Thanks, that helped a lot. Although I think the better solution would be to edit it a few lines earlier from:

`try {
      tailwindPackagePath = require.resolve('tailwindcss', { paths: [root] });
 }`

to this:

`try {
            tailwindPackagePath = require.resolve('@tailwindcss/postcss', { paths: [root] });
        }`

**Comment by EdKaim** [comment]
The easiest solution I found for the `ng test` issue is to just [rename `tailwind.config.js` to something else](https://github.com/SharpLogic/LightNap/issues/21), like `tailwind.config.lightnap.js`. Obviously update your references elsewhere, but it should work as long as the stuff you're testing doesn't rely on the contents of your tailwind.config settings.

I didn't dig into the details, but it seems that `ng test` behaves differently if there's a file called `tailwind.config.js` in the root, even if it's empty. Simply using a different name results in it not automatically invoking whatever that is. Apologies for not having a better explanation, but I was just looking to get unblocked and nothing else I tried worked.

**Comment by JOSERIDER** [comment]
> The easiest solution I found for the `ng test` issue is to just [rename `tailwind.config.js` to something else](https://github.com/SharpLogic/LightNap/issues/21), like `tailwind.config.lightnap.js`. Obviously update your references elsewhere, but it should work as long as the stuff you're testing doesn't rely on the contents of your tailwind.config settings.
> 
> I didn't dig into the details, but it seems that `ng test` behaves differently if there's a file called `tailwind.config.js` in the root, even if it's empty. Simply using a different name results in it not automatically invoking whatever that is. Apologies for not having a better explanation, but I was just looking to get unblocked and nothing else I tried worked.

Thats worked for me, only just remove the tailwind.config.js if you don't need.

**Comment by Ahmedhussein-92** [comment]
> [@HarelM](https://github.com/HarelM) — here are some potentially useful posts I've written about my experiences with Tailwind v4, Angular, and Angular Material:
> 
> * https://jits.dev/blog/migrating-angular-tailwind-v4/
> * https://jits.dev/blog/tailwind-v4-play-nice-with-angular-material/

@jits You my friend helped me tons with your posts. Thank you a lot! Much appreciated!

I was having an issue with conflicting styles between angular material 19 and tailwind v4 which was showing this silly border in the middle of some input components and some other unwanted behaviors. But your post helped me fix it. Thanks ton!

**Comment by HarelM** [comment]
The following is the PR I did to upgrade if that helps anyone: 
https://github.com/IsraelHikingMap/Site/pull/2161

**Comment by jansowinski** [comment]
I tried to run both tailwind 4 and 4.1 and failed. I am running angular 19.2 **custom-webpack builder** and primeng 19.1. Styles are just not applied. tailwind 3 works fine. 
With application builder everything runs smoothly - I checked this on other project.

**Comment by ganySA** [comment]
I have the exactly the same issue in angular 20.0.2. Angular complains about:

`Error: It looks like you're trying to use `tailwindcss` directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install `@tailwindcss/postcss` and update your PostCSS configuration.`

**Comment by yarikwest** [comment]
> I have the exactly the same issue in angular 20.0.2. Angular complains about:
> 
> `Error: It looks like you're trying to use `tailwindcss`directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install`@tailwindcss/postcss` and update your PostCSS configuration.`

That's not angular complains, but tailwindcss himself. So you need to do what error message says, install @tailwindcss/postcss

**Comment by ganySA** [comment]
> Can confirm. Tailwind v4 _mostly_ works, but [@apply](https://github.com/apply) inside component styles do not (they do in v3). CLI throws an error that the style doesn't exist

you can include @reference "tailwindcss" at the top of the component and the @apply will work, if you have a custom configuation you need to add that as a @reference too.

**Comment by clydin** [comment]
> > Can confirm. Tailwind v4 _mostly_ works, but [@apply](https://github.com/apply) inside component styles do not (they do in v3). CLI throws an error that the style doesn't exist
> 
> you can include [@reference](https://github.com/reference) "tailwindcss" at the top of the component and the [@apply](https://github.com/apply) will work, if you have a custom configuation you need to add that as a [@reference](https://github.com/reference) too.

This is a breaking change within Tailwind v4 itself.  As mentioned above, `@reference` is now required in components and affects other frameworks as well.  For more information: https://tailwindcss.com/docs/functions-and-directives#reference-directive

**Comment by noguespi** [comment]
Yes it works with `@reference`. The issue is you can only use `@reference` using relative path which is very cumbersome.

Also note you should use `@reference ../../[...]/your-styles.css` and not `@reference tailwind.css` [if you want your custom directive to be active](https://tailwindcss.com/docs/functions-and-directives#reference-directive).

When using `@reference` there are performance concerns with [vue, svelte astro](https://tailwindcss.com/docs/compatibility#vue-svelte-and-astro) and it probably apply to angular the same way:

> When using CSS modules, build tools like Vite, Parcel, and Turbopack process each CSS module separately. That means if you have 50 CSS modules in a project, Tailwind needs to run 50 separate times, which leads to much slower build times and a worse developer experience.


Not sure how big the impact is...

That's too bad because sometime it's just cleaner to do plain css using `@variant [screen-size]` than bloating a single line class attribute...

**Comment by ganySA** [comment]
I could not get that to work. when i reference my own style.scss it seems to fail. It does not adopt the @theme etc or anything else. This is really so much harder than it should be.

In a component i can only make @reference "tailwindcss" work

When i try reference my own main.scss or another which has my @use "tailwindcss" and then my @theme 

Then the whole thing breaks again - stuck!

Oh and to this is where AI fails - goes around in circles :) need a human here with a proper anwer.

**Comment by ganySA** [comment]
Interestingly if i use @use and include the file it works 100% but if i use @reference then it says cannot file my own personalised stylesheet.

**Comment by ganySA** [comment]
Aside from the above solution - the only thing that ultimately worked for me was to reload my tailwind configuration in the component, else my custom tailwind configuration is simply not available.

then reference tailwind. 

@use "../../../../../styles/tailwind-theme.scss"; 
@reference "tailwindcss";

**Comment by noguespi** [comment]
Beware that with tailwind v4 some errors will display an error message but won't break the build.  And it's very easy to miss these error messages.

See #30573

**Comment by hybrist** [comment]
@noguespi Thanks for also opening an upstream issue! https://github.com/tailwindlabs/tailwindcss/issues/18370

**Comment by noguespi** [comment]
I did open [another issue](https://github.com/tailwindlabs/tailwindcss/issues/18492) related to the mandatory use of relative path with `@reference`.


This issue combined with [errors not breaking the build](https://github.com/tailwindlabs/tailwindcss/issues/18370) make tailwind v4 really hard to use with Angular. 

If you forgot a `../` in your `@reference` and don't notice the error message you will ship bugged css.

**Comment by noguespi** [comment]
@wongjn suggested me a great workaround that seems to work:

> As a workaround, you can try using [subpath imports](https://nodejs.org/api/packages.html#subpath-imports):
> 
> ```js
> // package.json
> {
>   …
>   "imports": {
>     "#index.css": "./path/to/index.css"
>   },
>   …
> } 
> ```
> 
> ```css
> @reference "#index.css";
> ```

No more relative `@reference` !  See [discussion here](https://github.com/tailwindlabs/tailwindcss/discussions/18493#discussioncomment-13706707).

**Comment by weilinzung** [comment]
We are currently experiencing another style issue with the latest Angular builder. It seems to be something between Tailwind CSS 4 and Angular.

> The [latest update](https://github.com/angular/angular-cli/commit/49a09737d5412c302d09b40de198251bb99789d1) to the Angular CLI already includes support for “.postcssrc.json”.

See discussion here: https://github.com/tailwindlabs/tailwindcss/discussions/18520#discussion-8571707

---
Since the current Angular builder cannot enable production mode, we must enable it manually by setting `optimize: true`.
```json
{
  "plugins": {
    "@tailwindcss/postcss": {
      "optimize": true
    }
  }
}
```
> **Updated**: https://github.com/tailwindlabs/tailwindcss/discussions/18520#discussioncomment-13744050

**Comment by RockNHawk** [comment]
see also https://github.com/tailwindlabs/tailwindcss/discussions/17716

**Comment by ck4957** [comment]
> I am also facing same issue in angular. I am creating a library in angular and using storybook. I tried with both tailwind 3 and tailwind 4. mainly the issue is that angular is not able to trigger tailwind properly to create classes at runtime, but at compile time it's working. But major issue is that in default setting, only the tailwind default presets are coming, but tailwind is not able to find the classes, as it is not a running application but library using storybook. So now in default settings and configurations, tailwind is not even generating the classes, for that i found one fix that, if i specify the library path in tailwind config it works as like this: (https://github.com/ashutoshtyagi-xyz/tailwind3-test), there the issue was fixed when i gave the content path in config file,
> 
> But with the tailwind 4, as default there is no config file, but we can include it which i tried like this: https://github.com/ashutoshtyagi-xyz/tailwind4-test/tree/new-config
> 
> when i tried the same with tailwind4 as config file is allowed, i am getting error:
> 
> ```
> info => Starting manager..
> info => Starting preview..
> info Addon-docs: using MDX3
> info => Using implicit CSS loaders
> info => Using angular browser target options from "dsx:build"
> info => Using angular project with "tsConfig:/Users/atyagi/Documents/Personal/tailwind4-test/projects/dsx/.storybook/tsconfig.json"
> => Failed to build the preview
> Error: It looks like you're trying to use `tailwindcss` directly as a PostCSS plugin. The PostCSS plugin has moved to a separate package, so to continue using Tailwind CSS with PostCSS you'll need to install `@tailwindcss/postcss` and update your PostCSS configuration.
>     at ze (./node_modules/.pnpm/tailwindcss@4.1.3/node_modules/tailwindcss/dist/lib.js:36:1984)
>     at getStylesConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/tools/webpack/configs/styles.js:96:66)
>     at async Promise.all (index 1)
>     at async generateWebpackConfig (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:64:22)
>     at async generateBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:119:20)
>     at async generateI18nBrowserWebpackConfigFromContext (./node_modules/.pnpm/@angular-devkit+build-angular@18.2.18_@angular+compiler-cli@18.2.13_@angular+compiler@1_7feff540033c3bf15585118c558e63c5/node_modules/@angular-devkit/build-angular/src/utils/webpack-browser-config.js:70:20)
>     at async exports.getWebpackConfig (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/angular-cli-webpack.js:40:35)
>     at async Object.webpackFinal (./node_modules/.pnpm/@storybook+angular@8.6.12_2f2624659dd5fe59020cc76fcbd75413/node_modules/@storybook/angular/dist/server/framework-preset-angular-cli.js:20:27)
>     at async Object.webpack (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/presets/custom-webpack-preset.js:1:3280)
>     at async starter (./node_modules/.pnpm/@storybook+builder-webpack5@8.6.12_esbuild@0.23.1_storybook@8.6.12_typescript@5.5.4/node_modules/@storybook/builder-webpack5/dist/index.js:1:22249)
> ```
> 
> there is something from this line of code i think is not able to process the config file properly in tailwind4, that'll be due to [Angular code](https://github.com/angular/angular-cli/blob/807ba251c244f899c99769cc00c444051d3dd033/packages/angular_devkit/build_angular/src/tools/webpack/configs/styles.ts#L82-L98) that only integrates v3:
> 
> ```
>   const tailwindConfigPath = await findTailwindConfigurationFile(root, projectRoot);
>   if (tailwindConfigPath) {
>     let tailwindPackagePath;
>     try {
>       tailwindPackagePath = require.resolve('tailwindcss', { paths: [root] });
>     } catch {
>       …
>     }
>     if (tailwindPackagePath) {
>       extraPostcssPlugins.push(require(tailwindPackagePath)({ config: tailwindConfigPath }));
>     }
>   }
> ```
> 
> For v4, it should resolve @tailwindcss/postcss.
> 
> If any one has any workaround for this that will be useful. My target is to have a in-company design system using tailwind, a library that can be integrated in angular projects.

This solution worked for me with Angular v19 and custom-webpack solution
https://stackoverflow.com/questions/79632905/does-tailwindcss-v4-1-support-angular-v15

**Comment by noguespi** [comment]
Still 2 majors issues with tailwind 4 at the moment:

- Some kind of errors will be silently ignored and totally break your CSS. For example, missing a single `;` after a `@reference "#index.css"` will make the whole css file totally ignored without a single warning 🤯 it may be __partially__ fixed by [PR#18918](https://github.com/tailwindlabs/tailwindcss/pull/18918) (not released yet).

- And even if you are lucky to have warnings or erros reported, the build will complete succesfully and you will ship bugged code into production because of this [issue with PostCSS and angular](https://github.com/tailwindlabs/tailwindcss/issues/18370).

**Comment by clydin** [comment]
Those are issues with Tailwind v4 itself and are not related to Angular specifically.  Unfortunately, there are limited options for Angular to workaround those problems. Tailwind v4 does not fully propagate errors through the postcss system but rather prints them to the console. This prevents postcss and anything using postcss to be able to react to the errors since they are effectively invisible to the upstream systems.


As to general support within Angular, the upcoming v21 release will include Tailwind as a stylesheet system option within the `ng new` process.  Additionally, `ng add tailwindcss` will integrate Tailwind via `postcss` into an existing project and will attempt to augment both existing usage of `postcss` and Sass within that project.  You can try out the in-development versions of these features with the latest v21 prerelease (currently [21.0.0-next.4](https://github.com/angular/angular-cli/releases/tag/21.0.0-next.4)).

**Comment by noguespi** [comment]
I'm not sure it's a good idea to mix sass and Tailwind, it will probably lead to unexpected bugs and weird behavior because tailwind itself is already a preprocessor. It's already not working well in pure css with Angular so adding scss in the mix ...

Regarding [#18370](https://github.com/tailwindlabs/tailwindcss/issues/18370) it's not yet 100% clear if the issue is in Tailwind. Yes they are not propagating the errors in postcss plugin, but If I remember, they did it because Angular in watch mode (ng serve) would  [crash when tailwind throws errors](https://github.com/tailwindlabs/tailwindcss/pull/17754).

**Comment by clydin** [comment]
I'm not aware of that issue. The application build system within the Angular CLI specifically converts unhandled postcss errors into an error result object for the bundler.
However, Tailwind should ideally not be throwing exceptions for user code errors but rather reporting via postcss's error reporting mechansim where possible.

**Comment by rishabhmhjn** [comment]
Checking in to see if anyone got the `(.postcssrc|postcss.config).(js|mjs|cjs|ts|mts|cts)` files working?
`json` file has no issues

**Comment by burck1** [comment]
FYI for anyone interested. Sharing because it took me a while to figure out. I'm not associated with the Angular team nor the Tailwind team.

If you are using the `@angular-devkit/build-angular:browser` builder (webpack-based) and would like to upgrade to Tailwind v4, you will first need to upgrade to **Angular 20+**. Support for loading plugins from a PostCSS config file with the webpack builder(s) was added in [this commit](https://github.com/angular/angular-cli/commit/49a09737d5412c302d09b40de198251bb99789d1) and released in v20.something of the [@angular-devkit/build-angular](https://www.npmjs.com/package/@angular-devkit/build-angular) package.

If you can't update to Angular 20+, you'll probably need to stick with Tailwind v3 as the older versions of the [@angular-devkit/build-angular](https://www.npmjs.com/package/@angular-devkit/build-angular) package only supports loading the `tailwindcss` npm package as the postcss plugin (`require(tailwindPackagePath)({ config: tailwindConfigPath })`) when a [known](https://github.com/angular/angular-cli/blob/main/packages/angular/build/src/utils/postcss-configuration.ts#L21C7-L21C26) Tailwind config file is detected. This method of configuring tailwind is no longer supported in Tailwind v4.

**Comment by Kludgy4** [comment]
Are there plans to support the new faster tailwind build system that doesn't use postcss? It is slowing down my build :(

**Comment by quedicesebas** [comment]
> FYI for anyone interested. Sharing because it took me a while to figure out. I'm not associated with the Angular team nor the Tailwind team.
> 
> If you are using the `@angular-devkit/build-angular:browser` builder (webpack-based) and would like to upgrade to Tailwind v4, you will first need to upgrade to **Angular 20+**. Support for loading plugins from a PostCSS config file with the webpack builder(s) was added in [this commit](https://github.com/angular/angular-cli/commit/49a09737d5412c302d09b40de198251bb99789d1) and released in v20.something of the [@angular-devkit/build-angular](https://www.npmjs.com/package/@angular-devkit/build-angular) package.
> 
> If you can't update to Angular 20+, you'll probably need to stick with Tailwind v3 as the older versions of the [@angular-devkit/build-angular](https://www.npmjs.com/package/@angular-devkit/build-angular) package only supports loading the `tailwindcss` npm package as the postcss plugin (`require(tailwindPackagePath)({ config: tailwindConfigPath })`) when a [known](https://github.com/angular/angular-cli/blob/main/packages/angular/build/src/utils/postcss-configuration.ts#L21C7-L21C26) Tailwind config file is detected. This method of configuring tailwind is no longer supported in Tailwind v4.

So tailwind 4 it is supported form Angular 20?

**Comment by alan-agius4** [comment]
> So tailwind 4 it is supported form Angular 20?

Yes it is.
